eth/protocols/snap/sync.go

```diff
@@ -2899,3 +2900,76 @@ func (s *capacitySort) Swap(i, j int) {
 	s.ids[i], s.ids[j] = s.ids[j], s.ids[i]
 	s.caps[i], s.caps[j] = s.caps[j], s.caps[i]
 }
+
+// healRequestSort implements the Sort interface, allowing sorting trienode
+// heal requests, which is a prerequisite for merging storage-requests.
+type healRequestSort struct {
+	hashes []common.Hash
+	paths  []trie.SyncPath
+}
+
+func (t *healRequestSort) Len() int {
+	return len(t.hashes)
+}
+
+func (t *healRequestSort) Less(i, j int) bool {
+	a := t.paths[i]
+	b := t.paths[j]
+	switch bytes.Compare(a[0], b[0]) {
+	case -1:
+		return true
+	case 1:
+		return false
+	}
+	// identical first part
+	if len(a) < len(b) {
+		return true
+	}
+	if len(b) < len(a) {
+		return false
+	}
+	if len(a) == 2 {
+		return bytes.Compare(a[1], b[1]) < 0
+	}
+	return false
+}
+
+func (t *healRequestSort) Swap(i, j int) {
+	t.hashes[i], t.hashes[j] = t.hashes[j], t.hashes[i]
+	t.paths[i], t.paths[j] = t.paths[j], t.paths[i]
+}
+
+// Merge merges the pathsets, so that several storage requests concerning the
+// same account are merged into one, to reduce bandwidth.
+// OBS: This operation is moot if t has not first been sorted.
+func (t *healRequestSort) Merge() []TrieNodePathSet {
+	var nPathset []TrieNodePathSet
+	var last TrieNodePathSet
+
+	for _, path := range t.paths {
+		pSet := TrieNodePathSet([][]byte(path))
+		if len(path) == 1 {
+			nPathset = append(nPathset, pSet)
+			last = nil
+			continue
+		}
+		if last != nil && bytes.Equal(pSet[0], last[0]) {
+			// we can merge this into the last one
+			last = append(last, pSet[1])
+			nPathset[len(nPathset)-1] = last
+		} else {
+			nPathset = append(nPathset, pSet)
+			last = pSet
+		}
+	}
+	return nPathset
+}
+
+// sortByAccountPath takes hashes and paths, and sorts them. After that, it generates
+// the TrieNodePaths and merges paths which belongs to the same account path.
+func sortByAccountPath(hashes []common.Hash, paths []trie.SyncPath) ([]common.Hash, []trie.SyncPath, []TrieNodePathSet) {
+	n := &healRequestSort{hashes, paths}
+	sort.Sort(n)
```

Comment:
Isn't dangerous(or wrong) to reorder the requested node?

The order is generated by trie syncer which will always request children first then parent. If we break this order, the assumption is broken.

Reply:
NVM. In trie syncer it will maintain a `dep` counter for each requested trie node, to ensure all the children are persisted first. So it's OK and necessary to sort requests first.